text/0000-packages-as-optional-namespaces.md

```diff
@@ -0,0 +1,182 @@
+
+We currently prevent `foo-bar` and `foo_bar` from existing at the same time. We _could_ do this here as well, but it would only go in one direction: if `foo/bar` exists, neither `foo-bar` nor `foo_bar` will be allowed to be published. However, if `foo-bar` or `foo_bar` exist, we would choose to allow `foo/bar` to be published, because we don't want to limit the use of names within a crate namespace due to crates outside the namespace existing. This limits the "damage" to cases where someone pre-squats `foo-bar` before you publish `foo/bar`, and the damage can be mitigated by checking to see if such a clashing crate exists when publishing, if you actually care about this attack vector. There are some tradeoffs there that we would have to explore.
+
+One thing that could mitigate `foo/bar` mapping to the potentially ambiguous `foo_bar` is using something like `foo::crate::bar` or `~foo::bar` or `foo::/bar` in the import syntax.
+
+
+## Slow migration
+
+Existing projects wishing to use this may need to manually migrate. For example, `unic-langid` may become `unic/langid`, with the `unic` project maintaining `unic-langid` as a reexport crate with the same version number. Getting people to migrate might be a bit of work, and furthermore maintaining a reexport crate during the (potentially long) transition period will also be some work. Of course, there is no obligation to maintain a transition crate, but users will stop getting updates if you don't.
+
+A possible path forward is to enable people to register aliases, i.e. `unic-langid` is an alias for `unic/langid`.
+
+# Rationale and alternatives
+
+This change solves the ownership problem in a way that can be slowly transitioned to for most projects.
+
+## `foo::bar` on crates.io and in Rust
+
+While I cover a bunch of different separator choices below, I want to call out `foo::bar` in particular. If we went with `foo::bar`, we could have the same crate name in the Rust source and Cargo manifest. This would be _amazing_.
+
+Except, of course, crate `foo::bar` is ambiguous with module `bar` in crate `foo` (which might actually be a reexport of `foo::bar` in some cases).
+
+This can still be made to work, e.g. we could use `foo::crate::bar` to disambiguate, and encourage namespace-using crates to ensure that `mod bar` in crate `foo` either doesn't exist or is a reexport of crate `foo::bar`. I definitely want to see this discussed a bit more.
+
+## Separator choice
+
+A different separator might make more sense.
+
+We could continue to use `/` but also use `@`, i.e. have crates named `@foo/bar`. This is roughly what npm does and it seems to work. The `@` would not show up in source code, but would adequately disambiguate crates and features in Cargo.toml and in URLs.
```

Comment:
I am rather partial to this approach; it has a bunch of benefits:

 - It has a clear mapping to crates.io/docs.rs URLs
 - There's no ambiguity in feature specs
 - We _could_ use the same syntax directly in source (`@ <ident> /` triggers parsing as a namespaced path) though this might be too large a syntax change

Reply:
I do think it would be ideal for the same syntax to be used both in Cargo.toml and in source code. I'd like to suggest the usage of `:::` but I think it has the same issue that led to `...` becoming `..=`.

Can you expand on why `@foo::bar` might be too large a syntax change?

Reply:
Because it changes what counts as an identifier in Rust: it's a change at the _token_ level, which in turn affects proc macros and such.

Reply:
(it doesn't have to be at the token level tbh, but that does still complicate how paths work)

Reply:
Would be nice to have @rust-lang/lang input on this as a syntax change

(if this RFC starts going down this route I'll of course add them as an approving team)

Reply:
I'm hoping to post a PR tomorrow for adding `cargo-add` to cargo.

`@`: The PR will include the proposal to add the command which will also include a proposal to extend cargo's version syntax.  Currently, you can do `<name>:<version>` and we'll be proposing to add support for `<name>@<version>`.  I'll also be proposing to support this syntax in more commands.  The `:` syntax has had little visibility in cargo and with us making it more front-and-center, it got some extra scrutiny, especially when comparing to other `add` commands which led to the cargo team being on board with changing the syntax.

`/`: cargo-add is allowing both `cargo add path/to/crate` and `cargo add <crate-name>`.   The current logic is
- If the parameter exists as a path, use it as a path
- If the parameter contains a `/`, assume the user intended a path and error
- Else, treat it as a `<name>` or `<name>@<version>` and provide whatever error comes from that

I feel like supporting `/` in crate paths will make things ambiguous enough that we'd need to re-think how we specify dependencies in cargo-add.

Reply:
I think what @Manishearth means when he is saying large syntax change is about the rust namespaces accepting `@` and not cargo.

Reply:
what does `@` mean in npm's context?

Reply:
> what does `@` mean in npm's context?


The way npm does package names in a package.json is like `@scope/package@version`. 
Where scope is a existing organization on npm and `@version` a published version. 

Though they also allow you to use Github repos as shown in the image 
![image](https://user-images.githubusercontent.com/20520193/157639732-fb8ea587-b67a-4b76-a8fe-686a86b53183.png)

Reply:
please don't introduce new keywords. I like the idea of namespace, but please take a simple path, or don't implement it all until you find one. rust is already complex enough, I'm affraid we have to write things like this in the future: 

```
from crate<'a,T> import T::foo::<'a> where T: Pin<Box<dyn Crate>>> + 'a Namespace + Copy + Debug as @foo
```

Reply:
Using a character that is currently not possible within a crate name does avoid issues like a bunch of crates already being published with a name that want as a root. For example, if your org uses something like `foo_bar` and `foo_baz` but someone else already published foo, you can't use those names. But you could for something like `@foo/bar`.

That said, I see @ehuss's [comment below](https://github.com/rust-lang/rfcs/pull/3243#issuecomment-1816785092) that cargo has already closed on support for `::`. Does that effectively mean this RFC is decided?

Reply:
Its been approved by 2 of the 3 related teams and is just waiting on enough votes from the last team.